File: Windows/OfflineMessenger.Core/Protocol/PacketSerializer.cs
using System;
using System.IO;
using System.Buffers.Binary;

namespace OfflineMessenger.Core.Protocol;

public static class PacketSerializer
{
    public static byte[] Serialize(MessagePacket packet)
    {
        using var ms = new MemoryStream();
        using var bw = new BinaryWriter(ms);

        // 🔹 HEADER
        bw.Write((byte)packet.Type);
        bw.Write(GuidToBytes(packet.SessionId));
        bw.Write(GuidToBytes(packet.MessageId));

        var timestamp =
            new byte[8];

        BinaryPrimitives.WriteInt64BigEndian(
            timestamp,
            packet.Timestamp
        );

        bw.Write(timestamp);

        // Nonce
        var nonceLength = new byte[4];

        BinaryPrimitives.WriteInt32BigEndian(
            nonceLength,
            packet.Nonce.Length
        );

        bw.Write(nonceLength);
        bw.Write(packet.Nonce);


        // Tag
        var tagLength = new byte[4];

        BinaryPrimitives.WriteInt32BigEndian(
            tagLength,
            packet.Tag.Length
        );

        bw.Write(tagLength);
        bw.Write(packet.Tag);


        // Payload
        var payloadLength = new byte[4];

        BinaryPrimitives.WriteInt32BigEndian(
            payloadLength,
            packet.Payload.Length
        );

        bw.Write(payloadLength);
        bw.Write(packet.Payload);

        return ms.ToArray();
    }

    public static MessagePacket Deserialize(byte[] data)
    {
        using var ms = new MemoryStream(data);
        using var br = new BinaryReader(ms);

        var packet = new MessagePacket();

        // 🔹 HEADER
        packet.Type = (MessageType)br.ReadByte();
        packet.SessionId = BytesToGuid(
        br.ReadBytes(16)
            );
        packet.MessageId =
            BytesToGuid(
                br.ReadBytes(16)
            );


        packet.Timestamp =
            BinaryPrimitives.ReadInt64BigEndian(
                br.ReadBytes(8)
            );
        packet.Timestamp = br.ReadInt64();

        // 🔹 Nonce
        int nonceLen =
            BinaryPrimitives.ReadInt32BigEndian(
                br.ReadBytes(4)
            );

        packet.Nonce =
            br.ReadBytes(nonceLen);


        // 🔹 Tag
        int tagLen =
            BinaryPrimitives.ReadInt32BigEndian(
                br.ReadBytes(4)
            );

        packet.Tag =
            br.ReadBytes(tagLen);


        // 🔹 Payload
        int payloadLen =
            BinaryPrimitives.ReadInt32BigEndian(
                br.ReadBytes(4)
            );

        packet.Payload =
            br.ReadBytes(payloadLen);

        return packet;
    }

    public static byte[] SerializeHandshake(HandshakePacket packet)
    {
        using var ms = new MemoryStream();
        using var bw = new BinaryWriter(ms);

        bw.Write((byte)packet.Type);

        bw.Write(packet.PublicKey.Length);
        bw.Write(packet.PublicKey);

        return ms.ToArray();
    }

    public static HandshakePacket DeserializeHandshake(byte[] data)
    {
        using var ms = new MemoryStream(data);
        using var br = new BinaryReader(ms);

        var packet = new HandshakePacket();

        packet.Type = (MessageType)br.ReadByte();

        int len = br.ReadInt32();
        packet.PublicKey = br.ReadBytes(len);

        return packet;
    }

    private static byte[] GuidToBytes(Guid guid)
    {
        var result = new byte[16];

        var bytes = guid.ToByteArray();

        Buffer.BlockCopy(
            bytes,
            0,
            result,
            0,
            16
        );

        return result;
    }

    private static Guid BytesToGuid(byte[] bytes)
    {
        return new Guid(bytes);
    }
}

File: Android/app/src/main/java/protocol/PacketSerializer.kt
package com.offlinemessenger.android.protocol

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.util.UUID
import android.util.Log

object PacketSerializer {


    fun serialize(packet: MessagePacket): ByteArray {

        val output = ByteArrayOutputStream()
        val writer = DataOutputStream(output)


        writer.writeByte(
            packet.type.ordinal + 1
        )

        writer.write(
            uuidToBytes(packet.sessionId)
        )

        writer.write(
            uuidToBytes(packet.messageId)
        )

        writer.writeLong(
            packet.timestamp
        )


        writer.writeInt(
            packet.nonce.size
        )

        writer.write(
            packet.nonce
        )


        writer.writeInt(
            packet.tag.size
        )

        writer.write(
            packet.tag
        )


        writer.writeInt(
            packet.payload.size
        )

        writer.write(
            packet.payload
        )


        writer.flush()

        return output.toByteArray()
    }



    fun deserialize(data: ByteArray): MessagePacket {

        val input =
            DataInputStream(
                ByteArrayInputStream(data)
            )

        Log.d(
            "SERIALIZER",
            data.joinToString(
                prefix = "",
                separator = " "
            ) {
                "%02X".format(it)
            }
        )

        val packet = MessagePacket()


        packet.type =
            when (input.readByte().toInt()) {

                1 -> MessageType.HandshakeInit
                2 -> MessageType.HandshakeReply
                3 -> MessageType.Chat

                else ->
                    throw Exception(
                        "Unknown packet type"
                    )
            }


        packet.sessionId =
            bytesToUuid(
                readBytes(input, 16)
            )


        packet.messageId =
            bytesToUuid(
                readBytes(input, 16)
            )


        packet.timestamp =
            input.readLong()



        val nonceSize =
            input.readInt()

        packet.nonce =
            readBytes(
                input,
                nonceSize
            )



        val tagSize =
            input.readInt()

        packet.tag =
            readBytes(
                input,
                tagSize
            )



        val payloadSize =
            input.readInt()

        packet.payload =
            readBytes(
                input,
                payloadSize
            )


        return packet
    }



    private fun readBytes(
        input: DataInputStream,
        size: Int
    ): ByteArray {

        if (size < 0 || size > 10_000_000) {
            throw IllegalArgumentException(
                "Invalid packet size: $size"
            )
        }

        val buffer = ByteArray(size)

        input.readFully(buffer)

        return buffer
    }



    private fun uuidToBytes(
        uuid: UUID
    ): ByteArray {

        val bytes =
            ByteArray(16)

        val buffer =
            java.nio.ByteBuffer.wrap(bytes)


        buffer.putLong(
            uuid.mostSignificantBits
        )

        buffer.putLong(
            uuid.leastSignificantBits
        )


        return bytes
    }



    private fun bytesToUuid(
        bytes: ByteArray
    ): UUID {

        val buffer =
            java.nio.ByteBuffer.wrap(bytes)


        return UUID(
            buffer.long,
            buffer.long
        )
    }

    fun serializeHandshake(
        packet: HandshakePacket
    ): ByteArray {

        val output =
            ByteArrayOutputStream()

        val writer =
            DataOutputStream(output)


        writer.writeByte(
            packet.type.ordinal + 1
        )


        writer.writeInt(
            packet.publicKey.size
        )


        writer.write(
            packet.publicKey
        )


        writer.flush()


        return output.toByteArray()
    }
}